### ISSUE TITLE
Major Security Flaw

### ISSUE THREAD

**Issue by hhsnopek** [carrot/oauth2_proxy#1]
When you have a valid cookie present you're able to access any app regardless if you have permissions or not. 😳

Fix: Add permissions check when validating cookies

**Comment by BrandonRomano** [comment]
👀

**Comment by hhsnopek** [comment]
cc @roideuniverse - this is the issue that's top priority for the backend suite. Basically we just need to check the access for the user when we check access on each url request, tho this will up the usage of the permissions api, we might want to do some caching in the permissions api in RAM or something to hold a users permissions a certain amount of time (like 2 hours?)

**Comment by roideuniverse** [comment]
So, the first problem I faced was that when the cookie is available the `callback?state=/compass.carrotdashboard.com` was not available so we didn't had a app name. I added a hack to get the app name by using `req.Host` when `state` was not present.

This seems to have solved this bug. But this also introduces another problem.

When a user is authorized for an app, lets say `testapp.carrotdashboard.com` a valid cookie is created. 
Now when the user tries to access another app `compass.carrotdashboard.com` for which he is unauthorized, then `oauth2_proxy` removes that cookie. 

Now when the logged in user tries to access `testapp.carrotdashboard.com`, he is no longer logged in and is asked to login again. 

I am open to suggestions. We can either accept this behavior or I can modify the internal cookie logic.
